### ISSUE TITLE
404 Release Version web pages not available 

### ISSUE THREAD

**Issue by OmarJAH** [schemaorg/schemaorg#2805]
Hi there,

The releases page ( https://schema.org/docs/releases.html ) has links to each version of schema.org. 

The web pages for the versions are not working for at least some days already. e.g. https://schema.org/version/11.0/ or https://schema.org/version/7.0/

**Comment by RichardWallis** [comment]
A couple of releases ago we stopped shipping the data files for all the previous versions onto the live schema.org web site.  These files, for previous versions, are vary rarely accessed, and are available from the GitHub repository.

The links from the releases page need updating to account for this thus:

- [version 10.0](https://github.com/schemaorg/schemaorg/tree/main/data/releases/10.0) - (https://github.com/schemaorg/schemaorg/tree/main/data/releases/10.0)
- [version 7.01](https://github.com/schemaorg/schemaorg/tree/main/data/releases/7.01) - (https://github.com/schemaorg/schemaorg/tree/main/data/releases/7.01)  

The only exception to this is the '_current_' version which is always available from the link https://schema.org/version/latest

**Comment by danbri** [comment]
We should also make sure that previously publicised URLs like
/version/11.0/ are not 404.

**Comment by tetron** [comment]
What is the correct developer behavior here?  I have files that reference http://schema.org/version/9.0/schemaorg-current-http.rdf which are all broken now.

Previously they referenced "latest" but some time ago that URL changed.

So then I changed them to reference a specific version, but now those URLs are dead.

I don't understand why the most commonly used schema on the internet can't provide stable URIs.

**Comment by RichardWallis** [comment]
The default method of access to the vocabulary definition files for the most recently released version is via the ```version/latest``` populated URLs  as defined on the [developers page](https://schema.org/docs/developers.html#defs).

> eg. ```https://schema.org/version/latest/schemaorg-current-http.jsonld```

For performance and stability reasons, only the latest releases of these files are made available from the live schema.org web site.  Those wishing to access to those from previous versions can obtain them from the releases archive area of the GitHub repository: https://github.com/schemaorg/schemaorg/tree/main/data/releases

This issue needs finalising to confirm any outstanding incorrect links in the releases listing

**Comment by tetron** [comment]
If "files, for previous versions, are vary rarely accessed" I don't really follow why they need to be purged for "performance and stability".

So if I would like to identify a specific schema.org release (in a machine-readable way, to substitute a cached copy instead of the links that are currently 404) is there a preferred stable URI I should use?  Just a direct link to github?  A git hash?  Continue to use `http://schema.org/version/9.0/schemaorg-current-http.rdf` even though it doesn't resolve?

**Comment by RichardWallis** [comment]
> to substitute a cached copy instead of the links that are currently 404

Are you referring to 404 links within vocabulary definitions?

Currently this issue is only focussed on specific links within the releases.html page, if you are aware of others please add them to this issue.

For a version identifier I suggest using 11.0, 10,0, 7.01, etc.

These could then feature in the Github download URLs for previous versions if you need to construct them.

**Comment by tetron** [comment]
Let me explain my use case.

I have a file format to which users can add arbitrary metadata annotations.  Here's an example

https://github.com/common-workflow-library/bio-cwl-tools/blob/b7542dc3f4fed4c4b2cb8ba2e322798df4798aba/fastqc/fastqc_2.cwl#L174-L211

We recommend but do not enforce using schema.org annotations.  Because we can support annotations from arbitrary schemas, we support linking to the schema in use for a particular document, which is then automatically downloaded and used to do some basic validation of the user's annotations.

We previously were linking to `http://schema.org/version/latest/schema.rdf` but that URL changed last year.

So we replaced all of our references to `http://schema.org/version/9.0/schemaorg-current-http.rdf` on the idea that being version pinned, these URLs would be more stable, but now those URLs return 404.

We could use `https://schema.org/version/latest/schemaorg-current-http.rdf` but now I am not confident that it won't change again in the future.  Also, even if there is a 100% backwards compatibility guarantee, it is still easy to be in a situation where someone has used a newer term that is unrecognized in an older version of the schema, but it is impossible to know if you are in that situation if they both declare they are the "latest" schema.

What I want is an RDF URI patttern that recognizably identifies (a) the schema.org schema and (b) the particular version of the schema. Which the `http://schema.org/version/9.0/` prefix does.  I could continue to use that and substitute a cached copy of the schema, but it is unfortunate that it is no longer dereferencible.

If there's another way I am supposed to handle this, I'd like to know.

As a side note, there's also a w3id entry for schema.org, but the redirect hasn't been updated since 2014: https://github.com/perma-id/w3id.org/tree/master/schema.org

**Comment by mr-c** [comment]
If stable URLs at schema.org aren't an option, the I agree with @tetron that updating the w3id redirects is a good enough compromise.

We can send a PR to update those w3id.org redirects, if the project agrees to mention them in the documentation here.

**Comment by danbri** [comment]
We should put something sensible at http://schema.org/version/9.0/ etc.

Whether it should be a human-readable spec too, or just a wrapper pointing to the machine definitions ... --- any preferences here? It adds site codebase and content complexity to generate and maintain these as spec-like pages for humans, particularly when Github keeps an entire history.

**Comment by OmarJAH** [comment]
> Whether it should be a human-readable spec too, or just a wrapper pointing to the machine definitions ... --- any preferences here?

IMHO the best thing would be to have content negotiation: if any RDF format is requested in the HTTP-header, then the vocabulary file in that format should be returned ("all-layers" for <= 8.0 ; "schemaorg-all-http" for >= 9.0 ). If requested by a browser, a simple page with meta information (date, amount of classes, etc.) about that particular release and further links to machine-readable formats would be enough.

Personally, I queried these pages to make sure that a specific version number has been officially released since the versions file at https://raw.githubusercontent.com/schemaorg/schemaorg/latest/versions.json has already entries for the next release (e.g. "schemaversion": "12.0").

**Comment by danbri** [comment]
We are not going to do content negotiation.

It adds too much complexity and contradicts Schema.org’s general approach
to structured data publication.

(The newer JSON-LD spec also supports alternative approaches to
context-file publication now)

**Comment by stain** [comment]
Would be a good idea that a versioned structured schema for Web development remains available on the web with structure and versions.

If the current Web hosting is too unstable to provide reliable URIs then we should look for other alternatives - many good cloud providers out there.

**Comment by stain** [comment]
Reminding everyone of this old classic: [Cool URIs don't change](https://www.w3.org/Provider/Style/URI)

If a "live" and "archive" site is need to kept distinct, then with almost all hosting providers possible to do pattern-based or specific HTTP redirects. No content negotiation needed - this is about access to particular formats of particular versions of schema.org vocabulary.

**Comment by RichardWallis** [comment]
@tetron thank you for explaining your use case I can see where you are coming from.

Such a case however has not been previously considered or accounted for. 

Bringing the discussion back to the original issue raised here, there is some work on my schedule to fix 404 links in the releases.html documentation page.  I intend to move on to it a soon as my schedule frees up a bit.

Beyond then, and on the subject of stable and reliable URIs for the vocabulary description files (as described on the [developers page](https://schema.org/docs/developers.html#defs)), they will be available from two locations:

- The currently live version: ```https://schema.org/version/latest/{file name}```
- All versions: ```https://github.com/schemaorg/schemaorg/tree/main/data/releases/{version number}/{file name}```

Final note.  The fact that version 12 is already visible in the ```versions.json``` in the repository, albeit with a placeholder date (2021-XX-XX), is an indication that preparation is under way for the next release and at some stage be test deployed to the webschemas.org site.

**Comment by danbri** [comment]
It's not about the service provision, it's about deciding what would be useful content to put at those URLs. Their primary role is to serve as entry points identifying each release. Useful conventions for how to go from those URLs to actual machine definitions will likely evolve over time, so we are not going to commit to never changing what is found at the links. 

I think there's general agreement that serving 404s for URLs previously advertised as representing schema.org releases is simply a bug, and not worth discussing forever instead of fixing.

**Comment by RichardWallis** [comment]
Fix to links in releases.html, and onward link to associated archive area containing version specific download files, applied in PR #2815.

Fix will become live with next version release (V12.0)

**Comment by stain** [comment]
Thanks @RichardWallis - I notice https://schema.org/version/10.0/ etc. are still 404 and awaiting this fix - but #2815 does not seem to resolve this as it only changes to GitHub links?

**Comment by stain** [comment]
Which @id should be used for https://schema.org/schemaVersion ? I have assumed for now that https://schema.org/version/10.0/ will be restored as documented in that property.

**Comment by RichardWallis** [comment]
@stain You raise a good point.

Although the [schemaVersion](https://schema.org/schemaVersion) property, usable on [CreativeWork](https://schema.org/CreativeWork) and its subtypes, has a Schema.org specific example in its documentation, it is a general property that could identify the version of any schema associated with the entity being described - ether as a URL or a Text value. 

The value used in the example ```http://schema.org/version/2.0/``` somewhat dates the last time attention was focused on the description of this property by the community.  Version 2.0 was released in May 2015, when https based URIs were still just an interesting suggestion.

So to try to answer your question, against that background... 
It depends on what is available to data publishers and, we expect consumers of the data to do with it.  For some [Dataset](https://schema.org/Dataset)s a text value would be all that you can expect, for others a link to a descriptive html page, or a link to a download file.

If you were describing a specific release of the Schema.org vocabulary, in its current form as discussed at length above, there are three options:
1.  A text value "2.0", "7.04", "11.0", etc.
2. The URL to the Github location that contains the download files for a particular version.  eg:
     ```https://github.com/schemaorg/schemaorg/tree/main/data/releases/11.0```
3. The URL of one of those download files. eg:
     ```https://raw.githubusercontent.com/schemaorg/schemaorg/main/data/releases/11.0/schemaorg-current-https.ttl```

Pragmatically I would suggest that the first of these would be the simplest and future proof answer.

**Comment by danbri** [comment]
We should redirect http and https to the corresponding #v2.0 id on
/docs/releases.html and update the documentation accordingly

**Comment by RichardWallis** [comment]
We should update the description for [schemaVersion](https://schema.org/schemaVersion) property to not include that link, and expand it to indicate that is it is not specifically for a Schema.org schemaVersion.   I will raise an issue for that.

The recent changes that will appear with the V12.0  release should take care of the rest.

**Comment by RichardWallis** [comment]
Discussion around schema version identifiers moved into issue #2841

**Comment by gkellogg** [comment]
Adding a link checker, such as the [`lycheeverse/lychee-action@v1` GitHub action](https://github.com/lycheeverse/lychee-action) to the continuous integration step might be a good way to check the site for bad links. I don't do it any more, but I've used the following:

```
   - name: Link Checker
      uses: lycheeverse/lychee-action@v1
      with:
        args: -v index.html --exclude http://example.org/ --exclude https://w3c.github.io/rdf-star/rdf-star-cg-spec.html/
```

**Comment by RichardWallis** [comment]
Implemented in V12.0

**Comment by stain** [comment]
> Implemented in V12.0

https://schema.org/version/12.0/ is 404 Not Found now btw and so are later versions. Only https://schema.org/version/22.0/ works for some reason.
